### PR TITLE
Adjust implementation of set.add() method, add set comm tests

### LINKED ISSUE #17214: Set
### Summary of Problem

What behavior did you observe when encountering this issue?
When I use set to keep the frontier vertices in  breadth first search (BFS) algorithm, for a special line graph, the program with 2 locales will break down. However, When I use Domain or DistBag to keep the vertices, no errors happen.

The  delaunay benchmark will cause the same error too. 
https://sparse.tamu.edu/DIMACS10/delaunay_n17
 

### Steps to Reproduce

**Source Code:**
The BFS codes can be accessed here
https://github.com/Bader-Research/arkouda/blob/graph-multilocales/src/SegmentedMsg.chpl#L2357

```chapel
// Please provide source code that will reproduce the problem.
// You can insert your code inline if it's not too long.
// Otherwise, you can attach it as a file or provide a URL to it.
// To the extent possible, providing simplified programs demonstrating the
// problem will be appreciated.
// You can also replace this section with "Associated Future Test(s)" below.
```
**Execution command:**
The server side will start an arkouda like this
./arkouda_server -nl 2

the client side is like this
python bfs.py nodexx 5555

If an input file is required, include it as well. -->
The line graph is just like this 
0 1
1 2
2 3
3 4
4 5
5 6
...
999 1000

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
Adjust implementation of set.add() method, add set comm tests (#17346)

Resolves #17214.

Prior to this effort, the `set` type would segfault whenever a remote
`add()` call triggered a resize. Adding an `on this` block to the
implementation of `add()` would fix the issue, but also caused an
extra copy of the added element to be made (see #15808).

By using `Memory.Initialization.moveToValue()`, the added element can
be moved into the `on this` block without making an extra copy.

This was not the end of the copy woes. An extra copy was still being
made because copy elision was not triggering for local variables in
`if` or `else` blocks. This was resolved by #17321.

Add a comm test that is intended to trigger the segfault illustrated
in #17214.

Add a test capturing the number of copies for a single `add()` call.

Add a future illustrating that a user defined `operator==` fails to
resolve.

Adjust `--warn-unstable` to only warn about `include` statements
in user modules.

TESTING:

- [x] `ALL` on `linux64` when `COMM=none`
- [x] `ALL` on `linux64` when `COMM=gasnet`

Reviewed by @mppf. Thanks!

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
Hi @lydia-duncan does the change to `include` statements and `--warn-unstable` seem OK to you?

**Comment by lydia-duncan** [comment]
> Hi @lydia-duncan does the change to `include` statements and `--warn-unstable` seem OK to you?

Seems fine to me, thanks!

File: test/library/standard/Set/setCommTest.good [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+init int
+loc=4
+init int
+loc=5
+init int
+loc=6
+init int
+loc=7
+init int
+loc=8
```
Comment (mppf) [inline test/library/standard/Set/setCommTest.good:19]: can you do something to make this test output more understandable? it's kindof long and monolithic

File: test/modules/bradc/printModStuff/foo.good [hunk 2]
```diff
@@ -49,6 +49,9 @@ end of module search dirs
   $CHPL_HOME/modules/layouts/LayoutCS.chpl
   $CHPL_HOME/modules/dists/SparseBlockDist.chpl
   $CHPL_HOME/modules/dists/ReplicatedDist.chpl
+  $CHPL_HOME/modules/standard/Memory.chpl
+  $CHPL_HOME/modules/standard/Memory/Diagnostics.chpl
+  $CHPL_HOME/modules/standard/Memory/Initialization.chpl
```
Comment (mppf) [inline test/modules/bradc/printModStuff/foo.good:54]: I didn't think we were using Set in the included by default modules but I guess we are. Do you happen to know where?

File: modules/standard/Set.chpl [hunk 3]
```diff
@@ -143,12 +143,26 @@ module Set {
 
     // Returns true if the key was added to the hashtable.
     pragma "no doc"
-    proc _addElem(in elem: eltType): bool {
-      var (isFullSlot, idx) = _htb.findAvailableSlot(elem);
+    proc _addElem(pragma "no auto destroy" in elem: eltType): bool {
```
Comment (mppf) [inline modules/standard/Set.chpl:146]: It would be nice if we didn't need the pragma to do it. Could you make an issue about that (or something)?

**Review by mppf — APPROVED** [review]

File: modules/standard/Set.chpl [hunk 4]
```diff
@@ -143,12 +143,26 @@ module Set {
...
-      if isFullSlot then return false;
-      _htb.fillSlot(idx, elem, none);
-      return true;
+      var result = false;
+
+      on this {
+
+        // TODO: If we try to move onto this too early, and then call the
+        // 'findAvailableSlot' method on 'moved' instead of 'elem', we'll
+        // get lifetime errors in '.../Set/types/testNilableTuple.chpl'.
```
[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/Set.chpl:155]: I think maybe the comment could be this

```
// TODO: put `var moved = moveToValue(elem)` here and
// call findAvailableSlot(moved) to simplify things.
// As of this writing, that leads to lifetime errors in
// '.../Set/types/testNilableTuple.chpl'.

File: test/modules/bradc/printModStuff/foo.good (hunk 2 above)
Comment (dlongnecke-cray) [reply test/modules/bradc/printModStuff/foo.good:54]: I'm not sure where. @lydia-duncan mentioned something about it earlier?

Comment (lydia-duncan) [reply test/modules/bradc/printModStuff/foo.good:54]: It's not included by default but I believe it is still used in a default included module.  There's a Sung test that prints the module initialization order, I think that'll indicate where it is first used

File: modules/standard/Set.chpl (hunk 3 above)
Comment (dlongnecke-cray) [reply modules/standard/Set.chpl:146]: https://github.com/cray/chapel-private/issues/1796

File: test/modules/bradc/printModStuff/foo.good (hunk 2 above)
Comment (dlongnecke-cray) [reply test/modules/bradc/printModStuff/foo.good:54]: Looks like it is used in `standard/Random.chpl`, in the implementation of `_choiceUniform`.

**Comment by zhihuidu** [comment]
@dlongnecke-cray Thanks and I can run the previous tests well now.

**Pull request merged**